File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051216Z_claude_claude-sonnet-5_php-stdlib/lib/play_calendar.php
<?php
declare(strict_types=1);

// Routes: Play (campaign-scoped calendar the DM initializes once, advances in
// bounded day increments, and exposes to authenticated campaign members with
// deterministic weather)
// ---------------------------------------------------------------------------

const CALENDAR_SEASON_OFFSETS = ['spring' => 0, 'summer' => 1, 'autumn' => 2, 'winter' => 3];
const CALENDAR_WEATHER_MAP = [0 => 'clear', 1 => 'rain', 2 => 'wind', 3 => 'snow'];

function calendar_weather(int $day, string $season): string {
    $offset = CALENDAR_SEASON_OFFSETS[$season];
    return CALENDAR_WEATHER_MAP[($day + $offset) % 4];
}

function calendar_response(array $calendar): array {
    return [
        'day' => (int)$calendar['day'],
        'season' => $calendar['season'],
        'weather' => calendar_weather((int)$calendar['day'], $calendar['season']),
    ];
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/calendar$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    if ($actor['username'] !== $campaign['owner']) {
        forbidden('only the campaign dm may initialize the calendar');
    }

    $body = read_json_body();
    if ($body === null
        || !isset($body['day'], $body['season'])
        || !is_valid_int_range($body['day'], 1, PHP_INT_MAX)
        || !is_string($body['season'])
        || !array_key_exists($body['season'], CALENDAR_SEASON_OFFSETS)) {
        bad_request();
    }
    $day = (int)$body['day'];
    $season = $body['season'];

    $stmt = $db->prepare('SELECT day, season FROM play_campaign_calendars WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) !== false) {
        conflict('calendar is already initialized for this campaign');
    }

    $stmt = $db->prepare('INSERT INTO play_campaign_calendars (campaign_id, day, season) VALUES (?, ?, ?)');
    $stmt->execute([$campaignId, $day, $season]);

    send_json(calendar_response(['day' => $day, 'season' => $season]), 201);
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/calendar/advance$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    if ($actor['username'] !== $campaign['owner']) {
        forbidden('only the campaign dm may advance the calendar');
    }

    $body = read_json_body();
    if ($body === null
        || !isset($body['days'])
        || !is_valid_int_range($body['days'], 1, 30)) {
        bad_request();
    }
    $days = (int)$body['days'];

    $stmt = $db->prepare('SELECT day, season FROM play_campaign_calendars WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    $calendar = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($calendar === false) {
        not_found('calendar not initialized for this campaign');
    }

    $newDay = (int)$calendar['day'] + $days;

    $stmt = $db->prepare('UPDATE play_campaign_calendars SET day = ? WHERE campaign_id = ?');
    $stmt->execute([$newDay, $campaignId]);

    send_json(calendar_response(['day' => $newDay, 'season' => $calendar['season']]), 200);
}

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/calendar$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    $isDm = $actor['username'] === $campaign['owner'];
    if (!$isDm && !is_play_campaign_member($db, $campaignId, $actor['username'])) {
        forbidden('only the owner or a party member may view the calendar');
    }

    $stmt = $db->prepare('SELECT day, season FROM play_campaign_calendars WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    $calendar = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($calendar === false) {
        not_found('calendar not initialized for this campaign');
    }

    send_json(calendar_response($calendar), 200);
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T082348Z_claude_claude-sonnet-5_go-stdlib/play_calendar.go
package main

import (
	"encoding/json"
	"net/http"
)

// playCalendar is a campaign's day/season calendar, initialized once by the
// dm and advanced in bounded day increments.
type playCalendar struct {
	Day    int
	Season string
}

// seasonOffsets maps each valid season to its weather-computation offset.
var seasonOffsets = map[string]int{
	"spring": 0,
	"summer": 1,
	"autumn": 2,
	"winter": 3,
}

var weatherByIndex = []string{"clear", "rain", "wind", "snow"}

// playCalendarWeather derives the deterministic weather for a day/season.
func playCalendarWeather(day int, season string) string {
	idx := (day + seasonOffsets[season]) % 4
	return weatherByIndex[idx]
}

func playCalendarResponse(cal *playCalendar) map[string]interface{} {
	return map[string]interface{}{
		"day":     cal.Day,
		"season":  cal.Season,
		"weather": playCalendarWeather(cal.Day, cal.Season),
	}
}

// handlePlayCampaignCalendarSub routes the "calendar" and "calendar/advance"
// sub-paths of a play campaign. It returns false if rest does not name a
// calendar path, so the caller can fall through to its own routing.
func handlePlayCampaignCalendarSub(w http.ResponseWriter, r *http.Request, campaignID, rest string) bool {
	if rest == "calendar" {
		switch r.Method {
		case http.MethodPost:
			handleInitPlayCalendar(w, r, campaignID)
		case http.MethodGet:
			handleGetPlayCalendar(w, r, campaignID)
		default:
			writeError(w, http.StatusMethodNotAllowed, "method not allowed")
		}
		return true
	}
	if rest == "calendar/advance" {
		if r.Method != http.MethodPost {
			writeError(w, http.StatusMethodNotAllowed, "method not allowed")
			return true
		}
		handleAdvancePlayCalendar(w, r, campaignID)
		return true
	}
	return false
}

// handleInitPlayCalendar lets the campaign dm initialize the calendar once.
func handleInitPlayCalendar(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	var req struct {
		Day    *int   `json:"day"`
		Season string `json:"season"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.Day == nil || *req.Day < 1 {
		writeError(w, http.StatusBadRequest, "day must be an integer greater than or equal to 1")
		return
	}
	if _, validSeason := seasonOffsets[req.Season]; !validSeason {
		writeError(w, http.StatusBadRequest, "season must be one of spring, summer, autumn, or winter")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if c.Owner != username {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the dm may initialize the calendar")
		return
	}
	if c.Calendar != nil {
		playMu.Unlock()
		writeError(w, http.StatusConflict, "calendar is already initialized")
		return
	}

	c.Calendar = &playCalendar{Day: *req.Day, Season: req.Season}
	resp := playCalendarResponse(c.Calendar)
	playMu.Unlock()
	persistState()

	writeJSON(w, http.StatusCreated, resp)
}

// handleGetPlayCalendar returns the campaign calendar to any authenticated
// campaign member.
func handleGetPlayCalendar(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if !isPlayMember(c, username) {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the owner or a member may view the calendar")
		return
	}
	if c.Calendar == nil {
		playMu.Unlock()
		writeError(w, http.StatusNotFound, "calendar has not been initialized")
		return
	}
	resp := playCalendarResponse(c.Calendar)
	playMu.Unlock()

	writeJSON(w, http.StatusOK, resp)
}

// handleAdvancePlayCalendar lets the campaign dm advance the calendar by a
// bounded number of days.
func handleAdvancePlayCalendar(w http.ResponseWriter, r *http.Request, campaignID string) {
	username, ok := authenticatePlay(r)
	if !ok {
		writeError(w, http.StatusUnauthorized, "authentication required")
		return
	}

	var req struct {
		Days *int `json:"days"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.Days == nil || *req.Days < 1 || *req.Days > 30 {
		writeError(w, http.StatusBadRequest, "days must be an integer from 1 through 30")
		return
	}

	playMu.Lock()
	c, ok := requirePlayCampaign(w, campaignID)
	if !ok {
		return
	}
	if c.Owner != username {
		playMu.Unlock()
		writeError(w, http.StatusForbidden, "only the dm may advance the calendar")
		return
	}
	if c.Calendar == nil {
		playMu.Unlock()
		writeError(w, http.StatusNotFound, "calendar has not been initialized")
		return
	}

	c.Calendar.Day += *req.Days
	resp := playCalendarResponse(c.Calendar)
	playMu.Unlock()
	persistState()

	writeJSON(w, http.StatusOK, resp)
}
